src/QirRuntime/lib/Tracer/README.md

```diff
@@ -0,0 +1,177 @@
+| `void __quantum__qis__swap(%Qubit* %q1, %Qubit* %q2)` | See [Special handling of SWAP](#special-handling-of-swap) for details. |
+
+_Note on operation ids_: The client is responsible for using operation ids in a consistent manner. Operations with the
+ same id will be counted by the tracer as the _same_ operation, even accross invocations with different number of target
+ qubits or when different functors are applied.
+
+## Native backing of the extension methods ##
+
+The Resource Tracer will reuse qir-rt library as much as possible while extending it with the callbacks specified above.
+
+__Conditionals on measurements__: The Resource Tracer will execute LLVM IR's branching structures "as is", depending on
+ the values of the corresponding variables at runtime. To enable estimation of branches that depend on a measurement
+ result, the source Q# program will have to be authored in such a way that the Q# compiler will translate the
+ conditionals into `__quantum__qis__apply_if*` calls. The tracer will add operations from _both branches_ into the
+ layers it creates to compute the upper bound estimate.
+
+Nested conditionals, conditional measurements and conditional tracked operations will _not_ be supported.
+
+__Caching__ (lower priority): It might be a huge perf win if the Resource Tracer could cache statistics for repeated
+ computations. The Tracer will have an option to cache layering results per quantum module if the boundaries of modules
+ are treated as layering barriers.
+
+## Layering ##
+
+_Definition_: ___Time___ is an integer-valued function on all quantum operations in a program (gates, measurements,
+ qubits allocation/release). For each gate there is start and end times. For each qubit, there are times when the qubit
+ is allocated and released. Start time of a gate cannot be less than allocation time of any of the qubits the gate uses.
+ If two gates or measurements use the same qubit, one of the gates must have start time greater than or equal to the end
+ time of the other.
+
+A sequentially executed quantum program can be assigned a trivial time function, when all quantum operations have
+ duration of 1 and unique start times, ordered to match the flow of the program. Layering compresses the timeline by
+ assuming that some operations might be executed simultaneously while allowing for different operations to have various
+ durations.
+
+_Definition_: Provided a valid _time_ function for the program a ___layer of duration N at time T, denoted as L(T,N),___
+ is a subset of operations in the program such that all of these operations have start time greater or equal _T_ and
+ finish time less than _T + N_. The program is ___layered___ if all gates in it are partitioned into layers, that don't
+ overlap in time. The union of all qubits that are involved in operations of a given layer, will be denoted _Qubits(T,N)_.

```

Comment:
:-) I feel that a mathematicians' language is used here that often takes me some time to decipher (and the same concept typically can be explained simpler).
If I interpret the definition right, then 
_Layer_ is a set of operations that are executed (from start to completion) within certain period of time (within certain time frame/slot/window). That period of time starts at moment T (inclusive) and ends at moment T + N (exclusive).
If the interpretation is right, then for the term (_layer_, _layering_) to be more informative (for the newbies like me during the first reading) I would recommend to consider a term containing the word _time_, e.g. _time partition_ (_time group_), _time partitioning_ (_time grouping_). For instance, 
_Time partition_ is a set of operations that are executed (from start to completion) within certain period of time...
The program is _time-partitioned_ if all gates in it are grouped into non-overlapping time periods.

Reply:
Personally, I strongly favor single-word terminology ("layer") over multi-word ("time partitioning"). I understand that the definitions are dry: my goal was to describe the concepts as accurately as possible. Simplifying the language might feel good at first, but then you run into ambiguities, etc. To aid with comprehension I'll add diagrams.